File: Shooter/Floor.m
//
//  Floor.m
//  Shooter
//
//  Created by Sven A. Schmidt on 05/11/2013.
//  Copyright (c) 2013 feinstruktur. All rights reserved.
//

#import "Floor.h"

#import "FloorSegment.h"


@implementation Floor {
    SKNode *_segment;
}


+ (instancetype)floorAtHeight:(CGFloat)height inScene:(SKScene *)scene
{
    return [[self alloc] initAtHeight:height inScene:scene];
}


- (id)initAtHeight:(CGFloat)height inScene:(SKScene *)scene
{
    self = [super init];
    if (self) {
        // Because we will grow the segments we create them much higher than necessary and push them
        // beyond the bottom edge. That way we don't have to resize them and avoid flickering at the bottom.
        CGFloat negativeOffset = scene.size.height;
        _segment = [FloorSegment floorSegmentWithRect:CGRectMake(0, -negativeOffset,
                                                                 scene.size.width, height + negativeOffset)];
        [scene addChild:_segment];
    }
    return self;
}


@end

File: Shooter/Floor.swift
//
//  Floor.swift
//  Shooter
//
//  Created by Sven A. Schmidt on 18/06/2014.
//  Copyright (c) 2014 feinstruktur. All rights reserved.
//

import Foundation
import SpriteKit


let NumberOfSegments = 32

class Floor: NSObject {
    var segments: [FloorSegment] = []
    
    init(height: CGFloat, scene: SKScene) {
        assert(Int(scene.size.width) % NumberOfSegments == 0, "Width must be divisible by number of segments", file: "Floor.swift", line: 0)
        
        let segmentWidth = scene.size.width / CGFloat(NumberOfSegments)
        
        for i in 0..<NumberOfSegments {
            let x: CGFloat = CGFloat(i) * segmentWidth
            let y: CGFloat = 0
            let rect = CGRect(x: x, y: y, width: segmentWidth, height: height)
            var s = FloorSegment(rect: rect)
            s.previous = self.segments.last
            if s.previous {
                s.previous!.next = s
            }
            self.segments.append(s)
            scene.addChild(s)
        }
        super.init()
    }
    
    var maxHeight: CGFloat {
    get {
        var max: CGFloat = 0
        for seg in self.segments {
            max = seg.visibleHeight > max ? seg.visibleHeight : max
        }
        return max
    }
    }
}
